core/services/vrf/listener_v2.go

```diff
@@ -408,9 +413,38 @@ func (lsn *listenerV2) processRequestsPerSub(
 	)
 }
 
+func (lsn *listenerV2) estimateJuelsNeeded(
+	req *vrf_coordinator_v2.VRFCoordinatorV2RandomWordsRequested,
+	maxGasPriceWei *big.Int,
+) (*big.Int, error) {
+	// Don't use up too much time to get this info, it's not critical for operating vrf.
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	weiPerUnitLink, err := lsn.aggregator.LatestAnswer(&bind.CallOpts{Context: ctx})
+	if err != nil {
+		return nil, err
+	}
+	// NOTE: no need to sanity check this as this is for logging purposes only
+	// and should not be used to determine whether a user has enough funds in actuality,
+	// we should always simulate for that.
+	juelsNeeded := EstimateFeeJuels(
+		req.CallbackGasLimit,
+		maxGasPriceWei,
+		weiPerUnitLink,
+	)
+	return juelsNeeded, nil
+}
+
 // Here we use the pipeline to parse the log, generate a vrf response
 // then simulate the transaction at the max gas price to determine its maximum link cost.
-func (lsn *listenerV2) getMaxLinkForFulfillment(maxGasPrice *big.Int, req pendingRequest) (*big.Int, pipeline.Run, string, uint64, error) {
+func (lsn *listenerV2) getMaxLinkForFulfillment(maxGasPriceWei *big.Int, req pendingRequest) (*big.Int, pipeline.Run, string, uint64, error) {
+	// estimate how much juels are needed so that we can log it if the simulation fails.
+	juelsNeeded, err := lsn.estimateJuelsNeeded(req.req, maxGasPriceWei)
+	if err != nil {
+		// not critical, just log and continue
+		lsn.l.Debugw("unable to estimate juels needed for request, continuing anyway", "reqID", req.req.RequestId)
```

Comment:
That's a good question, that's up to us. My feeling is `Warn` is probably the highest we should go, since `Error` should probably be reserved for errors that cause fulfillments to fail (i.e simulation errored).

Reply:
Yeah SGTM